src/library/Numbers.nim

```diff
@@ -392,37 +393,48 @@ proc defineSymbols*() =
             #=======================================================
             push(newInteger(int(ceil(asFloat(x)))))
 
-    # TODO(Numbers/clamp) use a Range instead of min/max, as parameters
-    #  given how elegantly Range values are used, it would be better to use e.g a `range` 
-    #  Range type parameter for this one instead of a `max` and `min` parameter, so that we
-    #  we could do something along the lines of `clamp 2 1..3` (would return 2) or 
-    #  `clamp 4 1..3` (would return 3) or `clamp 0 1..3` (would return 1); same as before.
-    #  Also, should we reverse the order of the params, and have the range as the first one?
-    #  labels: library, enhancement
+
     builtin "clamp",
         alias       = unaliased, 
         op          = opNop,
         rule        = PrefixPrecedence,
         description = "force value within given range",
         args        = {
-            "number" : {Integer,Floating},
-            "min"    : {Integer,Floating},
-            "max"    : {Integer,Floating}
+            "number" : {Integer, Floating},
+            "range"  : {Range}
         },
         attrs       = NoAttrs,
-        returns     = {Integer,Floating},
+        returns     = {Integer, Floating},
         example     = """
-            clamp 2 1 3             ; 2
-            clamp 0 1 3             ; 1
-            clamp 4 1 3             ; 3
+            clamp 2 1..3                ; 2
+            clamp 0 1..3                ; 1
+            clamp 4 1..3                ; 3
+            clamp 4 3..1                ; 3
+            clamp 5 range.step: 2 1 5   ; 4
         """:
             #=======================================================
-            if x < y:
-                push(y)
-            elif x > z:
-                push(z)
+            if not y.rng.numeric:
+                panic RuntimeError, "incompatible types, range must be numeric"
+            
+            if x.kind == Integer:
+                if y.rng.forward:
+                    if x.i < y.rng.start: push(newInteger(y.rng.start))
+                    elif x.i > y.rng[y.rng.len, true].i: push(newInteger(y.rng[y.rng.len, true].i))
+                    else: push(x)
+                else:
+                    if x.i > y.rng.start: push(newInteger(y.rng.start))
+                    elif x.i < y.rng[y.rng.len, true].i: push(newInteger(y.rng[y.rng.len, true].i))
+                    else: push(x)
```

Comment:
Very nice job! 🚀 

Also, I like the handling of `.step:` (basically, I totally agree with your approach) and considering char ranges as invalid; it would be just too confusing... 😉 

```suggestion
                if (let minElem = y.rng.min()[1]; x.i < minElem.i): push(minElem)
                elif (let maxElem = y.rng.max()[1]; x.i > maxElem.i): push(maxElem)
                else: push(x)
```

Since we already have the [`min`](https://github.com/arturo-lang/arturo/blob/master/src/helpers/ranges.nim#L150-L158) / [`max`](https://github.com/arturo-lang/arturo/blob/master/src/helpers/ranges.nim#L160-L168) helper methods defined for Range values, and they do precisely that (return the min/max Value of a Range, regardless of whether it's a forward one or not; along with the index of that Value), why not just this? 😉 

Also, we can avoid getting the last/first elements twice, by using Nim's version of let-if syntax above. 😉  (Not the most intuitive way to write this sort of things, but what can we do...)

Reply:
Well, my bad. I had forgot about `min`/`max` functions. LOL!